L1Trigger/L1TTrackMatch/plugins/L1TrackSelectionProducer.cc

```diff
@@ -281,12 +283,22 @@ class L1TrackSelectionProducer : public edm::global::EDProducer<> {
         : deltaZMaxEtaBounds_(cfg.template getParameter<double>("deltaZMaxEtaBounds")),
           deltaZMax_(cfg.template getParameter<double>("deltaZMax")) {}
     bool operator()(const L1Track& t, const l1t::VertexWord& v) const {
+      TTTrack_TrackWord::tanl_t etaEmulationBits = t.getTanlWord();
+      ap_fixed<TrackBitWidths::kEtaSize, TrackBitWidths::kEtaMagSize> etaEmulation;
+      etaEmulation.V = etaEmulationBits.range();
       size_t etaIndex =
-          std::lower_bound(deltaZMaxEtaBounds_.begin(), deltaZMaxEtaBounds_.end(), std::abs(t.momentum().eta())) -
+          std::upper_bound(deltaZMaxEtaBounds_.begin(), deltaZMaxEtaBounds_.end(), std::abs(etaEmulation.to_double())) -
           deltaZMaxEtaBounds_.begin() - 1;
       if (etaIndex > deltaZMax_.size() - 1)
         etaIndex = deltaZMax_.size() - 1;
-      return std::abs(v.z0() - t.getZ0()) <= deltaZMax_[etaIndex];
+      l1t::VertexWord::vtxz0_t fixedTkZ0 = t.undigitizeSignedValue(
+          t.getZ0Bits(), TTTrack_TrackWord::TrackBitWidths::kZ0Size, TTTrack_TrackWord::stepZ0, 0.0);
+
+      ap_uint<TrackBitWidths::kPtSize> ptEmulationBits = t.getTrackWord()(
+          TTTrack_TrackWord::TrackBitLocations::kRinvMSB - 1, TTTrack_TrackWord::TrackBitLocations::kRinvLSB);
+      ap_ufixed<TrackBitWidths::kPtSize, TrackBitWidths::kPtMagSize> ptEmulation;
+      ptEmulation.V = ptEmulationBits.range();
```

Comment:
Where are these `ptEmulationBits` and `ptEmulation` actually used in this method?

Reply:
@aperloff Can you please answer this question?

Reply:
The information comes in as an N bit ap_uint, but it's really representing the charge and pt. The Nth bit is the charge and bits N-1 down to 0 are the pt. So the N-1 LSBs must be converted into an ap_ufixed. `ptEmulationBits` is the original set of bits in an ap_uint. `ptEmulation` is the more useful ap_fixed.

Reply:
Sorry to insist but, while I understood the _what_ thanks to your explanation, I still don't see the _where_. Unless there's some magic I don't see, `ptEmulation` doesn't participate to the `return` (while it does is some other operators [above](https://github.com/cms-sw/cmssw/blob/48524d14c45f86df515aa35fbefacbfa3424ca47/L1Trigger/L1TTrackMatch/plugins/L1TrackSelectionProducer.cc#L111)) 

```cpp 
return std::abs(v.z0() - fixedTkZ0.to_double()) <= deltaZMax_[etaIndex];
```

We only `v` coming from the inputs and `fixedTkZ0`, that makes sense since, as far as I have understood, this method checks the compatibility between a vertex and the track vertex.

Reply:
@AdrianoDee It's used here: https://github.com/cms-sw/cmssw/blob/48524d14c45f86df515aa35fbefacbfa3424ca47/L1Trigger/L1TTrackMatch/plugins/L1TrackSelectionProducer.cc#L583
That Selector is defined as part of an AND selector, and the instantiated AND selector is used here in L583.
Hope that clears things up.

Reply:
```suggestion
```

Reply:
This should be directly committable.